test/distributed/test_c10d_nccl.py

```diff
@@ -420,6 +420,47 @@ def allgather_base(output_t, input_t):
             # fails the check because the dtype is different
             allgather_base(output_t, tensor)
 
+    @requires_nccl()
+    def test_gather_ops(self):
+        store = c10d.FileStore(self.file.name, self.world_size)
+        pg = c10d.ProcessGroupNCCL(store, self.rank, self.world_size)
+
+        def gather(output_t, input_t, rootRank):
+            opts = c10d.GatherOptions()
+            opts.rootRank = rootRank
+            if rootRank == self.rank:
+                work = pg.gather(output_t, input_t, opts)
+            else:
+                work = pg.gather([], input_t, opts)
+            work.wait()
+
+        # init input
+        tensors = []
+        for i in range(self.num_gpus):
+            tensors.append(torch.tensor([self.rank]).cuda(i))
+
+        # init golden output
+        golden_ts = [[] for _ in range(self.num_gpus)]
+        for idx, ls in enumerate(golden_ts):
+            for _ in range(self.num_gpus):
+                ls.append(torch.tensor([idx]).cuda(idx))
+
+        # init output
+        output_ts = [[] for _ in range(self.num_gpus)]
+        for idx, ls in enumerate(output_ts):
+            for _ in range(self.world_size * self.num_gpus):
+                ls.append(torch.tensor([0]).cuda(idx))
+
+        gather(output_ts, tensors, self.rank)
```

Comment:
I think it is valuable to test the underlying ProcessGroup API, but that API is not public to PyTorch users. We should test this API more thoroughly: https://github.com/pytorch/pytorch/blob/master/torch/distributed/distributed_c10d.py#L2126 because it is what users use.

Reply:
We should test the public API as well, as I mentioned in my comment above^

Reply:
@pritamdamania87  @lishanYang I think there are already pretty comprehensive tests for the public API for scatter/gather here: https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/distributed/distributed_test.py#L2635 (similar tests exist for `gather` in `distributed_c10d`). 

Seems like all we need to do to enable that coverage of the public API is remove the skip decorators for NCCL.